forge-game/src/main/java/forge/game/ability/effects/DiscoverEffect.java

```diff
@@ -0,0 +1,175 @@
+package forge.game.ability.effects;
+
+import com.google.common.base.Predicates;
+import com.google.common.collect.Iterables;
+import forge.game.Game;
+import forge.game.ability.AbilityKey;
+import forge.game.ability.AbilityUtils;
+import forge.game.ability.SpellAbilityEffect;
+import forge.game.card.Card;
+import forge.game.card.CardCollection;
+import forge.game.card.CardLists;
+import forge.game.card.CardZoneTable;
+import forge.game.cost.CostDiscard;
+import forge.game.cost.CostPart;
+import forge.game.cost.CostReveal;
+import forge.game.player.Player;
+import forge.game.player.PlayerController;
+import forge.game.spellability.LandAbility;
+import forge.game.spellability.SpellAbility;
+import forge.game.trigger.TriggerType;
+import forge.game.zone.PlayerZone;
+import forge.game.zone.Zone;
+import forge.game.zone.ZoneType;
+import forge.util.CardTranslation;
+import forge.util.Localizer;
+
+import java.util.List;
+import java.util.Map;
+
+public class DiscoverEffect extends SpellAbilityEffect {
+
+    @Override
+    protected String getStackDescription(SpellAbility sa) {
+
+        return sa.getActivatingPlayer() + " discovers " + sa.getParamOrDefault("Num", "1") + ".";
+    }
+
+    @Override
+    public void resolve(SpellAbility sa) {
+        final Card host = sa.getHostCard();
+        final Game game = host.getGame();
+        final Player p = sa.getActivatingPlayer();
+
+        // Exile cards from the top of your library until you exile a nonland card with <N> mana value or less.
+        final int num = AbilityUtils.calculateAmount(host, sa.getParamOrDefault("Num", "1"), sa);
+
+        if (p == null || !p.isInGame()) return;
+
+        Card found = null;
+        CardCollection exiled = new CardCollection();
+        CardCollection rest = new CardCollection();
+
+        final PlayerZone library = p.getZone(ZoneType.Library);
+
+        for (int i = 0; i < library.size(); i++) {
+            final Card c = library.get(i);
+            exiled.add(c);
+            if (!c.isLand() && c.getCMC() <= num) {
+                found = c;
+                if (sa.hasParam("RememberDiscovered"))
+                    host.addRemembered(c);
+                break;
+            } else {
+                rest.add(c);
+            }
+        }
+
+        if (exiled.size() > 0) {
+            game.getAction().reveal(exiled, p, false);
+        }
+
+        changeZone(exiled, ZoneType.Exile, game, sa);
+
+        // Cast it without paying its mana cost or put it into your hand.
+        if (found != null) {
+            String prompt = Localizer.getInstance().getMessage("lblDiscoverChoice",
+                            CardTranslation.getTranslatedName(found.getName()));
+            final Zone origin = found.getZone();
+            final boolean play = p.getController().chooseBinary(sa, prompt,
```

Comment:
I still don't know if Binary Choice is a good idea there

wouldn't it be better to use Confirm Style for Casting?

@Agetian your opinion for this? right now the AI would try to cast it and then maybe fail later?